assignment3.cpp

```diff
@@ -0,0 +1,196 @@
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <vector>
+#include <string>
+#include <set>
+
+class Dictionary{
+	std::vector<std::string> dictionary;
+public:
+	Dictionary(const std::string &filemame){
+		std::ifstream file (filemame);
+  		if (file.is_open()) {
+ 			std::string line;
+    		while (std::getline (file,line)) {
+      			dictionary.push_back(line);
+    		}
+    		file.close();
+  		}
+	}
+	bool isPrefix(const std::string &word) const {
+		bool is_prefix = false;
+		for(int i = 0; i<dictionary.size() && !is_prefix; i++){
+			is_prefix = true;
+			for(int j = 0; j<word.size() && is_prefix; j++){
+				if(j<dictionary[i].size()){
+					is_prefix = dictionary[i][j] == word[j];
+				} else {
+					is_prefix=false;
+				} 
+			}
+		}
+		return is_prefix;
+
+	}
+	bool isWord(const std::string &word) const{
+		bool is_word = false;
+		for(int i = 0; i<dictionary.size() && !is_word; i++){
+			is_word = dictionary[i].compare(word) == 0;
+		}
+		return is_word;
+	}
+};
+
+class WordSearch{
+private:
+	typedef std::pair<int, int> slot;
+	
+	std::string getString(char **grid, int rows, int columns, 
+		                  std::vector<slot> &path){
+		std::string word;
+		for(int i=0; i<path.size(); i++){
+
+			word.push_back(grid[path[i].first][path[i].second]);
+		}
+		return word;
+	}
+
+	void findWord(char **grid, int rows, int columns, const Dictionary &dict, 
+				  std::vector<slot> &path, std::set<std::string> &solution){
+
+		std::string word = getString(grid,rows,columns,path);
+		if(!dict.isPrefix(word)) return;
+
+		if(dict.isWord(word)){
+			solution.insert(word); 
+		} 
+
+		slot current_pos = path[path.size()-1];
+
+		int init_row = current_pos.first-1;
+		int end_row = current_pos.first+1;
+		int init_col = current_pos.second-1;
+		int end_col = current_pos.second+1;
+
+		init_col = init_col<0?0:init_col;
```

Comment:
A question of taste, but some prefer to use `max` for this (and `min` in the other case)